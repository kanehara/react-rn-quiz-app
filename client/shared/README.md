# The shared module for the trivia clients
A module housing common JS code between the web and React Native clients.

## Contains
* Redux actions, reducers, getters
* Apollo client used in redux layer to fetch data
* styled components theme