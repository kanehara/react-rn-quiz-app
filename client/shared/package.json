{
  "name": "shared",
  "version": "0.0.1",
  "scripts": {
    "test": "jest"
  },
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/preset-env": "^7.3.1",
    "babel-jest": "^24.0.0",
    "jest": "^24.0.0",
    "unfetch": "^4.0.1"
  }
}
