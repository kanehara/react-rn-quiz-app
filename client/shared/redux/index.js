import { createStore, applyMiddleware, compose } from 'redux'
import thunk from 'redux-thunk'
import { quizReducer } from './quiz'
import { persistStore, persistReducer } from 'redux-persist'
import storage from 'redux-persist/lib/storage'

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ 
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) 
    : compose

const persistedReducer = persistReducer({
  key: 'root',
  storage
}, quizReducer)

const store = createStore(
  persistedReducer,
  composeEnhancers(
    applyMiddleware(thunk)
  )
)

const persistor = persistStore(store)

export {
  store,
  persistor
}
