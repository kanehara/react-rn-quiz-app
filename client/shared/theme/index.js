import theme from './theme'
import { ThemeProvider as T } from 'styled-components'
import React from 'react'

const ThemeProvider = ({children}) => <T theme={theme}>{children}</T>

export {
  ThemeProvider
}
