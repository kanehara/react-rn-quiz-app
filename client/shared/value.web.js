export default 'value in web2'
