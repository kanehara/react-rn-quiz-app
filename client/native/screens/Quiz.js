import React from 'react'
import {H1, H3, H5} from 'native/components/Header'
import View, {ScreenView} from 'native/components/View'
import {TrueButton, FalseButton} from 'native/components/Button'
import { connect } from 'react-redux'
import { actions, getters } from 'shared/redux/quiz'
import styled from 'styled-components'

const Card = styled(View)`
  padding: 25px 15px;
  background-color: white;
  border-radius: 15px;
  border: 1px solid #e5e5e5;
  /* box-shadow: 0px 5px 18px 5px #e5e5e5; */
  align-items: center;
  justify-content: center;
  width: 95%;
`

class QuizScreen extends React.Component {
  get question () {
    return this.props.questions[this.props.questionIndex] || {}
  }

  componentDidUpdate() {
    if (this.props.finished) {
      this.props.navigation.navigate('results')
    } else if (!this.question) {
      this.props.navigation.navigate('home')
    }
  }

  handleAnswer = answer => {
    this.props.setAnswer(answer)
    this.props.navigation.navigate({
      routeName: 'quiz',
      params: { index: this.props.questionIndex + 1 }
    })
  }

  render() {
    return (
      <ScreenView>
        <Card>
          <H3>{this.question.category}</H3>
          <H1>{this.question.question}</H1>
          <H5>{this.props.questionIndex + 1} of {this.props.questions.length}</H5>
          <TrueButton onPress={() => this.handleAnswer('True')} title='True' />
          <FalseButton onPress={() => this.handleAnswer('False')} title='False' />
        </Card>
      </ScreenView>
    );
  }
}

export default connect(
  state => ({
    finished: getters.finished(state),
    questions: state.questions,
    questionIndex: state.questionIndex
  }),
  dispatch => ({
    setAnswer: answer => dispatch(actions.setAnswer(answer))
  })
)(QuizScreen)
