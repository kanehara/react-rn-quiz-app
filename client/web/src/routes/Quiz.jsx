import React from 'react'
import { Ray, Flex } from "rayout"
import { connect } from 'react-redux'
import { actions, getters } from 'shared/redux/quiz'
import {Redirect} from 'react-router-dom'
import {TrueButton as _TrueButton, FalseButton} from '../components/Button'
import styled from 'styled-components'

const TrueButton = styled(_TrueButton)`
  margin-bottom: 35px;
`

const Card = styled.div`
  padding: 25px;
  background-color: white;
  border-radius: 15px;
  box-shadow: 0px 5px 18px 5px #e5e5e5;
  width: 95%;

  @media (min-width: 600px) {
    width: 80%;
  }
  
  @media (min-width: 1000px) {
    width: 50%;
  }
`

class Quiz extends React.Component {
  get question () {
    return this.props.questions[this.props.questionIndex]
  }

  render () {
    return this.props.finished
      ? <Redirect to='/results' />
      : this.question
        ? (
          <Card display='flex'>
            <h3>{this.question.category}</h3>
            <Ray minHeight='175px' marginBottom='20px'>
              <h1>{this.question.question}</h1>
              <h5>{this.props.questionIndex + 1} of {this.props.questions.length}</h5>
            </Ray>
            <Flex flexDirection='column'>
              <TrueButton onClick={() => this.props.setAnswer('True')}>True</TrueButton>
              <FalseButton onClick={() => this.props.setAnswer('False')}>False</FalseButton>
            </Flex>
          </Card>
        )
        : (
          <Redirect to='/' />
        )
  }
}

export default connect(
  state => ({
    finished: getters.finished(state),
    questions: state.questions,
    questionIndex: state.questionIndex
  }),
  dispatch => ({
    setAnswer: answer => dispatch(actions.setAnswer(answer))
  })
)(Quiz)
