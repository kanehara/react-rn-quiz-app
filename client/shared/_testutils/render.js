import React from 'react'
import { ThemeProvider } from 'styled-components'
import theme from 'shared/theme'
import { Provider } from 'react-redux' 
import { store } from 'shared/redux'
import renderer from 'react-test-renderer'

export const render = jsx => renderer.create(
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      {jsx}
    </ThemeProvider>
  </Provider>
)

