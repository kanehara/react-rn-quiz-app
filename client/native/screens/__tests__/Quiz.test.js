import {Quiz} from '../Quiz'
import React from 'react'
import { render } from 'shared/_testutils'
import {TouchableHighlight} from 'react-native'

describe('Quiz', () => {
  const getMockNavigation = () => ({
    navigate: jest.fn()
  })

  it('renders', () => {
    expect(render(<Quiz />)).toMatchSnapshot()
  })
  
  it('sets answer and navigates on to next question on answer', () => {
    const mockNav = getMockNavigation()
    const setAnswerMock = jest.fn()
    const instance = render(
      <Quiz navigation={mockNav} questionIndex={0} setAnswer={setAnswerMock} />
    )
    const buttons = instance.root.findAllByType(TouchableHighlight)
    buttons[0].props.onPress()
    expect(setAnswerMock).toHaveBeenCalledWith('True')
    expect(mockNav.navigate).toHaveBeenCalledWith({
      routeName: 'quiz',
      params: { index: 1 }
    })
    
    buttons[1].props.onPress()
    expect(setAnswerMock).toHaveBeenCalledWith('False')
    expect(mockNav.navigate).toHaveBeenCalledWith({
      routeName: 'quiz',
      params: { index: 1 }
    })
  })
})
