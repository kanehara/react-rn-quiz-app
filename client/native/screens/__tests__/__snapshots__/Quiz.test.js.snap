// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz renders 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5faf4",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
        "paddingBottom": 40,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 40,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "#e5e5e5",
          "borderRadius": 15,
          "borderStyle": "solid",
          "borderWidth": 1,
          "justifyContent": "center",
          "paddingBottom": 25,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 25,
          "width": "95%",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#292F36",
            "fontSize": 18,
            "fontWeight": "bold",
            "marginBottom": 10,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#292F36",
            "fontSize": 28,
            "fontWeight": "bold",
            "marginBottom": 16,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#292F36",
            "fontSize": 12,
            "fontWeight": "bold",
            "marginBottom": 6,
            "marginLeft": 0,
            "marginRight": 0,
            "marginTop": 4,
            "textAlign": "center",
          },
        ]
      }
    >
      NaN
       of 
      0
    </Text>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#c5e0de",
            "borderRadius": 10,
            "borderStyle": "solid",
            "borderWidth": 1,
            "justifyContent": "center",
            "marginBottom": 10,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 10,
            "paddingBottom": 5,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 5,
            "width": "100%",
          },
          Object {
            "backgroundColor": "#7095ff",
          },
        ]
      }
    >
      <View
        style={null}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#292F36",
                "fontSize": 16,
                "fontWeight": "700",
                "marginBottom": 4,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 4,
              },
            ]
          }
        >
          True
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderColor": "#c5e0de",
            "borderRadius": 10,
            "borderStyle": "solid",
            "borderWidth": 1,
            "justifyContent": "center",
            "marginBottom": 10,
            "marginLeft": 10,
            "marginRight": 10,
            "marginTop": 10,
            "paddingBottom": 5,
            "paddingLeft": 10,
            "paddingRight": 10,
            "paddingTop": 5,
            "width": "100%",
          },
          Object {
            "backgroundColor": "#ff5757",
          },
        ]
      }
    >
      <View
        style={null}
      >
        <Text
          style={
            Array [
              Object {
                "color": "#292F36",
                "fontSize": 16,
                "fontWeight": "700",
                "marginBottom": 4,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 4,
              },
            ]
          }
        >
          False
        </Text>
      </View>
    </View>
  </View>
</View>
`;
